File: server/web/routes/weather.py
"""
Weather route — fetches current conditions from the Open-Meteo API (no API key required).

Location is configured via env vars:
  WEATHER_LAT  (float, default 12.9716 = Bengaluru)
  WEATHER_LON  (float, default 77.5946 = Bengaluru)
  WEATHER_CITY (string, default "Bengaluru")

Returns:
  { temperature, feels_like, condition, high, low, humidity, wind_kph, aqi, city }

AQI is fetched from the Open-Meteo Air Quality API (no key required).
"""

import logging
import os
from typing import Optional

import httpx
from fastapi import APIRouter, Depends

from auth import get_optional_user

logger = logging.getLogger("meetingbox.weather")

router = APIRouter()

_WMO_CODES: dict[int, str] = {
    0: "Clear sky",
    1: "Mainly clear",
    2: "Partly cloudy",
    3: "Overcast",
    45: "Foggy",
    48: "Icy fog",
    51: "Light drizzle",
    53: "Drizzle",
    55: "Heavy drizzle",
    61: "Light rain",
    63: "Rain",
    65: "Heavy rain",
    71: "Light snow",
    73: "Snow",
    75: "Heavy snow",
    80: "Rain showers",
    81: "Rain showers",
    82: "Violent showers",
    95: "Thunderstorm",
    96: "Thunderstorm with hail",
    99: "Thunderstorm with heavy hail",
}


def _aqi_label(us_aqi: Optional[float]) -> Optional[str]:
    if us_aqi is None:
        return None
    v = int(us_aqi)
    if v <= 50:
        return "Good"
    if v <= 100:
        return "Moderate"
    if v <= 150:
        return "Unhealthy (Sensitive)"
    if v <= 200:
        return "Unhealthy"
    if v <= 300:
        return "Very Unhealthy"
    return "Hazardous"


@router.get("/weather")
async def get_weather(
    _current_user: Optional[dict] = Depends(get_optional_user),
) -> dict:
    lat = float(os.getenv("WEATHER_LAT", "12.9716"))
    lon = float(os.getenv("WEATHER_LON", "77.5946"))
    city = os.getenv("WEATHER_CITY", "Bengaluru")

    weather_url = (
        "https://api.open-meteo.com/v1/forecast"
        f"?latitude={lat}&longitude={lon}"
        "&current=temperature_2m,apparent_temperature,weathercode,relative_humidity_2m,wind_speed_10m"
        "&daily=temperature_2m_max,temperature_2m_min"
        "&timezone=auto&forecast_days=1"
    )
    aqi_url = (
        "https://air-quality-api.open-meteo.com/v1/air-quality"
        f"?latitude={lat}&longitude={lon}"
        "&current=us_aqi"
    )

    result: dict = {
        "city": city,
        "temperature": None,
        "feels_like": None,
        "condition": "Unknown",
        "high": None,
        "low": None,
        "humidity": None,
        "wind_kph": None,
        "aqi": None,
        "aqi_label": None,
    }

    async with httpx.AsyncClient(timeout=8.0) as client:
        try:
            resp = await client.get(weather_url)
            resp.raise_for_status()
            data = resp.json()
            curr = data.get("current", {})
            daily = data.get("daily", {})

            temp = curr.get("temperature_2m")
            result["temperature"] = round(temp) if temp is not None else None

            feels = curr.get("apparent_temperature")
            result["feels_like"] = round(feels) if feels is not None else None

            code = curr.get("weathercode")
            result["condition"] = _WMO_CODES.get(int(code), "Unknown") if code is not None else "Unknown"

            highs = daily.get("temperature_2m_max", [])
            lows = daily.get("temperature_2m_min", [])
            result["high"] = round(highs[0]) if highs else None
            result["low"] = round(lows[0]) if lows else None

            hum = curr.get("relative_humidity_2m")
            result["humidity"] = int(hum) if hum is not None else None

            wind = curr.get("wind_speed_10m")
            result["wind_kph"] = round(wind, 1) if wind is not None else None
        except Exception as exc:
            logger.warning("Weather fetch failed: %s", exc)

        try:
            aqi_resp = await client.get(aqi_url)
            aqi_resp.raise_for_status()
            aqi_data = aqi_resp.json()
            us_aqi = aqi_data.get("current", {}).get("us_aqi")
            result["aqi"] = int(us_aqi) if us_aqi is not None else None
            result["aqi_label"] = _aqi_label(us_aqi)
        except Exception as exc:
            logger.warning("AQI fetch failed: %s", exc)

    return result

File: frontend/src/api/weather.ts
export interface WeatherData {
  city: string
  temperature: number | null
  feels_like: number | null
  condition: string
  high: number | null
  low: number | null
  humidity: number | null
  wind_kph: number | null
  aqi: number | null
  aqi_label: string | null
}

const DEFAULT_WEATHER_CITY = 'Bengaluru'
const DEFAULT_WEATHER_LAT = 12.9716
const DEFAULT_WEATHER_LON = 77.5946
const WEATHER_REQUEST_TIMEOUT_MS = 8_000

const WMO_CODES: Record<number, string> = {
  0: 'Clear sky',
  1: 'Mainly clear',
  2: 'Partly cloudy',
  3: 'Overcast',
  45: 'Foggy',
  48: 'Icy fog',
  51: 'Light drizzle',
  53: 'Drizzle',
  55: 'Heavy drizzle',
  61: 'Light rain',
  63: 'Rain',
  65: 'Heavy rain',
  71: 'Light snow',
  73: 'Snow',
  75: 'Heavy snow',
  80: 'Rain showers',
  81: 'Rain showers',
  82: 'Violent showers',
  95: 'Thunderstorm',
  96: 'Thunderstorm with hail',
  99: 'Thunderstorm with heavy hail',
}

interface OpenMeteoForecast {
  current?: {
    temperature_2m?: number
    apparent_temperature?: number
    weathercode?: number
    relative_humidity_2m?: number
    wind_speed_10m?: number
  }
  daily?: {
    temperature_2m_max?: number[]
    temperature_2m_min?: number[]
  }
}

interface OpenMeteoAirQuality {
  current?: {
    us_aqi?: number
  }
}

interface IpLocation {
  city?: string
  region?: string
  country_name?: string
  latitude?: number | string
  longitude?: number | string
}

interface WeatherLocation {
  city: string
  lat: number
  lon: number
}

let ipLocationPromise: Promise<WeatherLocation> | null = null

function readNumberEnv(value: string | undefined, fallback: number): number {
  if (!value?.trim()) return fallback
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : fallback
}

function readOptionalNumberEnv(value: string | undefined): number | null {
  if (!value?.trim()) return null
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : null
}

function aqiLabel(usAqi: number | null): string | null {
  if (usAqi == null) return null
  if (usAqi <= 50) return 'Good'
  if (usAqi <= 100) return 'Moderate'
  if (usAqi <= 150) return 'Unhealthy (Sensitive)'
  if (usAqi <= 200) return 'Unhealthy'
  if (usAqi <= 300) return 'Very Unhealthy'
  return 'Hazardous'
}

async function fetchJson<T>(url: string, timeoutMs = WEATHER_REQUEST_TIMEOUT_MS): Promise<T> {
  const controller = new AbortController()
  const timeout = window.setTimeout(() => controller.abort(), timeoutMs)
  try {
    const response = await fetch(url, { signal: controller.signal })
    if (!response.ok) {
      throw new Error(`Weather request failed: ${response.status}`)
    }
    return response.json() as Promise<T>
  } finally {
    window.clearTimeout(timeout)
  }
}

function fallbackLocation(): WeatherLocation {
  return {
    city: import.meta.env.VITE_WEATHER_CITY?.trim() || DEFAULT_WEATHER_CITY,
    lat: readNumberEnv(import.meta.env.VITE_WEATHER_LAT, DEFAULT_WEATHER_LAT),
    lon: readNumberEnv(import.meta.env.VITE_WEATHER_LON, DEFAULT_WEATHER_LON),
  }
}

async function detectLocationByIp(): Promise<WeatherLocation> {
  const fixedLat = readOptionalNumberEnv(import.meta.env.VITE_WEATHER_LAT)
  const fixedLon = readOptionalNumberEnv(import.meta.env.VITE_WEATHER_LON)
  if (fixedLat != null && fixedLon != null) {
    return fallbackLocation()
  }

  if (!ipLocationPromise) {
    ipLocationPromise = fetchJson<IpLocation>('https://ipapi.co/json/', 5_000)
      .then((data) => {
        const lat = typeof data.latitude === 'string' ? Number(data.latitude) : data.latitude
        const lon = typeof data.longitude === 'string' ? Number(data.longitude) : data.longitude
        if (!Number.isFinite(lat) || !Number.isFinite(lon)) {
          return fallbackLocation()
        }

        return {
          city: data.city?.trim() || data.region?.trim() || data.country_name?.trim() || fallbackLocation().city,
          lat: lat as number,
          lon: lon as number,
        }
      })
      .catch(() => fallbackLocation())
  }

  return ipLocationPromise
}

export async function fetchWeather(): Promise<WeatherData> {
  const { lat, lon, city } = await detectLocationByIp()

  const forecastParams = new URLSearchParams({
    latitude: String(lat),
    longitude: String(lon),
    current: 'temperature_2m,apparent_temperature,weathercode,relative_humidity_2m,wind_speed_10m',
    daily: 'temperature_2m_max,temperature_2m_min',
    timezone: 'auto',
    forecast_days: '1',
  })
  const aqiParams = new URLSearchParams({
    latitude: String(lat),
    longitude: String(lon),
    current: 'us_aqi',
  })

  const result: WeatherData = {
    city,
    temperature: null,
    feels_like: null,
    condition: 'Unknown',
    high: null,
    low: null,
    humidity: null,
    wind_kph: null,
    aqi: null,
    aqi_label: null,
  }

  const [forecast, airQuality] = await Promise.allSettled([
    fetchJson<OpenMeteoForecast>(`https://api.open-meteo.com/v1/forecast?${forecastParams}`),
    fetchJson<OpenMeteoAirQuality>(`https://air-quality-api.open-meteo.com/v1/air-quality?${aqiParams}`),
  ])

  if (forecast.status === 'fulfilled') {
    const current = forecast.value.current ?? {}
    const daily = forecast.value.daily ?? {}

    result.temperature = current.temperature_2m != null ? Math.round(current.temperature_2m) : null
    result.feels_like = current.apparent_temperature != null ? Math.round(current.apparent_temperature) : null
    result.condition = current.weathercode != null ? WMO_CODES[current.weathercode] ?? 'Unknown' : 'Unknown'
    result.high = daily.temperature_2m_max?.[0] != null ? Math.round(daily.temperature_2m_max[0]) : null
    result.low = daily.temperature_2m_min?.[0] != null ? Math.round(daily.temperature_2m_min[0]) : null
    result.humidity = current.relative_humidity_2m != null ? Math.round(current.relative_humidity_2m) : null
    result.wind_kph = current.wind_speed_10m != null ? Math.round(current.wind_speed_10m * 10) / 10 : null
  }

  if (airQuality.status === 'fulfilled') {
    const usAqi = airQuality.value.current?.us_aqi
    result.aqi = usAqi != null ? Math.round(usAqi) : null
    result.aqi_label = aqiLabel(result.aqi)
  }

  return result
}
